addons/website_sale/models/website_snippet_filter.py

```diff
@@ -37,3 +37,60 @@ def _get_hardcoded_sample(self, model):
                 # merge definitions
             samples = merged
         return samples
+
+    def _get_products(self, website, product_source, limit, context):
+        handler = getattr(self, '_get_products_%s' % product_source, self._get_products_category)
+        products = handler(website, limit, context)
+        return products
+
+    def _get_products_category(self, website, limit, context):
+        search_domain = context.get('search_domain')
+        domain = [('website_published', '=', True)] + website.website_domain() + (search_domain or [])
+        return self.env['product.product'].search(domain, limit=limit)
+
+    def _get_products_recently_viewed(self, website, limit, context):
+        products = []
+        visitor = self.env['website.visitor']._get_visitor_from_request()
+        if visitor:
+            excluded_products = website.sale_get_order().mapped('order_line.product_id.id')
+            tracked_products = self.env['website.track'].sudo().read_group(
+                [('visitor_id', '=', visitor.id), ('product_id', '!=', False), ('product_id.website_published', '=', True), ('product_id', 'not in', excluded_products)],
+                ['product_id', 'visit_datetime:max'], ['product_id'], limit=limit, orderby='visit_datetime DESC')
+            products_ids = [product['product_id'][0] for product in tracked_products]
+            if products_ids:
+                products = self.env['product.product'].with_context(display_default_code=False).browse(products_ids[:limit])
+        return products
+
+    def _get_products_recently_sold_with(self, website, limit, context):
+        products = []
+        current_product_id = context.get('params').get('productId')
+        if current_product_id:
+            current_product_id = int(current_product_id)
+            sale_order = self.env['sale.order'].sudo().search([
+                ('state', 'in', ('sale', 'done')), ('order_line.product_id', '=', current_product_id)
+            ], limit=1, order='date_order DESC')
```

Comment:
maybe a limit >... 4/8 ? to have more products and be less intrusif about the last order that you could know.